### DISCUSSION TITLE
The use of getFieldState

### DISCUSSION THREAD

**Discussion by Moshyfawn** [react-hook-form/react-hook-form#7638] [Q&A] [answered]
I'm at a bit of a loss with coming up with real world examples where I'd use[ the new `getFieldState` form method](https://react-hook-form.com/api/useform/getfieldstate/#main), especially, with the fact that I also need to subscribe to [some form state](https://react-hook-form.com/api/useform/formstate) in order to be able to see an input field state update.
From [the RHF `getFieldState` docs section](https://react-hook-form.com/api/useform/getfieldstate/#main) 
> formState will need to subscribed.

Also, I see some unexpected _(to me)_ form field state behavior. It looks like the field state doesn't get updated.. form time to time? Even though, I've subscribed to the `form.isDirty` state.
See [this CSB](https://codesandbox.io/s/quirky-tesla-eo1x0?file=/src/App.tsx:411-414) for example.
1. Open [the CSB](https://codesandbox.io/s/get-field-state-example-cyz1x)
2. Notice `form.isDirty` state set to `false` and the First Name field is has it's states `invalid`, `isDirty`, `isTouched` set to their default `false` state.
3. Enter any non-empty string into the First Name field
4. Unfocus _(blur)_ the First Name input
4. Notice `form.isDirty`  is set to `true` due to the input's value change. `isDirty`, `isTouched` of the First Name input are set to `false`, even though the input's value has been modified and the input has been unfocused _(blured)_ after the the value change.

https://user-images.githubusercontent.com/16290753/150760901-2b351c46-4a40-45a9-a147-1d1fe1b1d7ad.mp4

**Comment by bluebill1049** [comment, marked as answer]
hey @Moshyfawn can you save your codesandbox?

**Reply by Moshyfawn** [reply]
Looks like CSB bugged out: the preview page was up to date, but the code has been reset lol :P
Updated and saved now. There's [a new URL](https://codesandbox.io/s/get-field-state-example-cyz1x) _(updated in the initial post as well)_

**Reply by bluebill1049** [reply]
`isDirty` is not set to `true`. it's due to `dirtyFields` is not subscribed.

one of the primary reasons to introduce this API is to try to close the gap with V7 and v8, as we want to deprecate `NestedValue`https://react-hook-form.com/ts#NestedValue type in the current v7 and try to leverage API like `getFieldState` instead. The nature of complex input field native primary vs object/array alike field type.

**Reply by bluebill1049** [reply]
> with the fact that I also need to subscribe to some form state in order to be able to see an input field state update.

Yea, I feel you on this one. However, we don't want to enable by default with global state/form level state update and give users the option to follow what they are subscribed to and also at which level.

**Reply by Moshyfawn** [reply]
Yup-yup, I understand the reasoning behind having to subscribe to the "right" form state, it totally makes sense for the way RHF is managing it. Having said that, the only application I found for the new API so far is to track the field's `invalid` state and it also seems far stretched of a use, since I can just enable native form validation with `noValidate` on the form element to get those `:valid`, `:invalid` CSS selectors for styling. All the other `isTouched`, `isDirty` don't make sense to me in terms of use for the UI modification or whatnot.
_Don't get me wrong, I'm not saying the API is bad. Just trying to understand where I can use the method, so it helps me_

**Reply by bluebill1049** [reply]
I agree above all. (Great feedback) The API itself was to align with what `useController` returns for `fieldState` so it's consistent through the library. I think this API will make more sense when you have a nested object/array-like data type field. As we all know the type 

```tsx
type Data = {
  // is this a group of fields or a single nested field
  guessWhatIam: {
    test: string,
    test1: number
  }
}

getFieldState() // will help you with the typing nightmare especially in the next major version
```

Personally, I would treat it as a field-level type-safe API and of course with some utility that some may or may not find useful. At the same time, I am more than welcome for potential improvement on the API as well.

**Reply by bluebill1049** [reply]
> All the other isTouched, isDirty don't make sense to me in terms of use for the UI modification or whatnot.

There is some good usage for form state as above, which is not supported by native which I hope one day we will.

- `isTouched` you only want to show validation error after the user is focused and blurs the input.
- `isDirty` you can indicate a UI tip indicating the input has been modified, similar to an unsaved file in our IDE.

**Reply by bluebill1049** [reply]
> isDirty is not set to true. it's due to dirtyFields is not subscribed.

I am addressing this part in the doc as well.

**Reply by bluebill1049** [reply]
<img width="697" alt="Screen Shot 2022-01-24 at 10 15 36 pm" src="https://user-images.githubusercontent.com/10513364/150773097-81b40b9c-4ac6-49ea-9bd1-8ae8dec20749.png">

**Reply by Moshyfawn** [reply]
> The API itself was to align with what useController returns for fieldState so it's consistent through the library

This point does make sense to me. Especially having an object with multiple fields constructing the whole value.

> isTouched you only want to show validation error after the user is focused and blurs the input.

Isn't it the `mode: "isTouched"` behavior?

> isDirty you can indicate a UI tip indicating the input has been modified, similar to an unsaved file in our IDE.

Huh, wouldn't `formState.isDirty` be appropriate for the use-case? You can check for `fieldState.isDirty` and highlight the modified value field, tho, I don't think that's a common pattern.

Not trying to bash all your points, @bluebill1049, btw. Just thinking "out-loud" :P 

> I am addressing this part in the doc as well.

This is a very good point, actually! 'cuz I was thinking that I should subscribe to the corresponding form state on the back of my mind, but it's a unique behavior / pattern, IMHO. So, seeing it clearly stated in the docs makes a lot of sense!

**Reply by bluebill1049** [reply]
> Isn't it the mode: "isTouched" behavior?

yes, correct. Auto mode vs manual mode like a car lol.

> Huh, wouldn't formState.isDirty be appropriate for the use-case? You can check for fieldState.isDirty and highlight the modified value field, tho, I don't think that's a common pattern.

yea, correct. Back to the original issue with the type issue with a nested object/array-like field, we are trying to be avoided with `NestedValue` type, so you don't need to apply this Generic around the field value instead use this function.

```tsx
type Form = {
  test: NestedValue<{xxxx}>
} 
formState.dirtyFields.test // without Nestedvalue this would generate type error, so it doesn't know it's a field or not, however nested value create other issues as well :(

// to 
type Form = {
  test: {xxxx}
} 
getFieldState('test')
```


> Not trying to bash all your points, @bluebill1049, btw. Just thinking "out-loud" :P

not to worry at all, I really enjoy this sort of discussion and you are a "pro" level. I wish I had this chat with you early on, I will make sure I do next time. 👍
